### PR TITLE
feat(core): example prediction markets using oa

### PR DESCRIPTION
**Pull request by Reinis-FRP**
Signed-off-by: Reinis Martinsons <reinis@umaproject.org>

**Motivation**

Demonstrate how to build prediction market on top of Optimistic Asserter.

**Summary**

Implements simple prediction market where outcome and payouts are resolved through Optimistic Asserter.

**Details**

Prediction Market contract allows to initialize prediction markets each having a pair of binary outcome tokens. Anyone can mint and burn the same amount of paired outcome tokens for the default payout currency. Trading of outcome tokens is outside the scope of this contract. Anyone can assert 3 possible outcomes (outcome 1, outcome 2 or split) that is verified through Optimistic Asserter. If the assertion is resolved true then holders of outcome tokens can settle them for the payout currency based on resolved market outcome.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            assertedOutcomeId == keccak256(market.p1Name) ||
+                assertedOutcomeId == keccak256(market.p2Name) ||
+                assertedOutcomeId == keccak256(p3Name),
+            "Invalid asserted outcome"
+        );
+
+        market.assertedOutcomeId = assertedOutcomeId;
+        uint256 minimumBond = oa.getMinimumBond(address(currency));
+        uint256 bond = market.requiredBond > minimumBond ? market.requiredBond : minimumBond;
+        bytes memory claim =
```
Comment (chrismaree) [inline packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol:95]: ^

File: packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    mapping(bytes32 => Market) public markets; // Maps marketId to Market struct.
+
+    mapping(bytes32 => AssertedMarket) public assertedMarkets; // Maps assertionId to AssertedMarket.
+
+    FinderInterface public immutable finder; // UMA protocol Finder used to discover other protocol contracts.
+    IERC20 public immutable currency; // Currency used for all prediction markets.
+    OptimisticAssertorInterface public immutable oa;
+    uint256 public constant assertionLiveness = 7200; // 2 hours.
+    bytes32 public immutable defaultIdentifier; // Identifier used for all prediction markets.
+    bytes public constant splitOutcome = "Unknown"; // Name of the split outcome.
```
Comment (md0x) [inline packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol:47]: Maybe we should rename this variable:
`splitOutcome` as I think it works as splitOutcome but also as an unresolvable/unknown etc...
Maybe `unresolvable` or `unknown`?

Comment (Reinis-FRP) [reply packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol]: updated to `unresolvable`

File: packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        delete assertedMarkets[assertionId];
+    }
+
+    // Dispute callback does nothing.
+    function assertionDisputed(bytes32 assertionId) public {}
+
+    // Mints pair of tokens representing the value of outcome1 and outcome2. Trading of outcome tokens is outside of the
+    // scope of this contract. The caller must approve this contract to spend the currency tokens.
+    function createOutcomeTokens(bytes32 marketId, uint256 tokensToCreate) public {
+        Market storage market = markets[marketId];
```
Comment (md0x) [inline packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol:185]: Should we block the creation of tokens if the market has been resolved?

Comment (md0x) [reply packages/core/contracts/optimistic-assertor/implementation/examples/PredictionMarket.sol:185]: Resolving this comment as it doesn't add any value

**Review by md0x — APPROVED** [review]
LGTM!

**Pull request merged**